### PR TITLE
`Programming exercise`: Fix inactive test cases being assigned to wrong task

### PR DESCRIPTION
**Pull request by simon-hng**
#### General

- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.cit.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Changes affecting Programming Exercises
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 1 (Atlassian Suite).
- [x] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab).

### Motivation and Context

There are some use cases where instructors use inactive test cases. 
To be consistent with the new additional of task based grading we fix this edge case to also work based on tasks.

### Description

Fixes an edge case where inactive test cases are no longer associated with their task

### Steps for Testing

Prerequisites:
- 1 Instructor

1. Create an exercise with the default settings.
1. Go to the exercise grading view, take a look at the view and remember the visible tasks/test cases or make a screenshot
1. Go to the exercise management and click on `Edit in editor`
1. Select the test repository and delete some test cases, making them inactive (e.g. behavioral)
1. Go back to the grading view and see if the test cases are still in their tasks

#### Performance Review
- [x] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [x] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ProgrammingExerciseTaskService | 88.77% | ✅              |

### CODE REVIEW

**Review by DominikRemo — DISMISSED** [review]
Tested in testing session on ts2. Works as expected 🚀

**Review by frederik-kt — APPROVED** [review]
Tested on TS2 in testing session. Commenting out the test case resulted in it only being shown when toggling the "Show inactive test cases" switch.

**Review by rriyaldhi — APPROVED** [review]
Tested on TS2 and works as expected.

File: src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseTaskService.java [hunk 1]
```diff
@@ -121,13 +121,16 @@ public List<ProgrammingExerciseTask> getSortedTasks(ProgrammingExercise exercise
...
     /**
-     * Gets all tasks of an exercise
+     * Gets all tasks of an exercise excluding inactive test cases
      *
      * @param exerciseId of the programming exercise
      * @return Set of all tasks and its test cases
      */
-    public Set<ProgrammingExerciseTask> getTasks(long exerciseId) {
-        return programmingExerciseTaskRepository.findByExerciseIdWithTestCaseAndSolutionEntriesElseThrow(exerciseId);
+    public Set<ProgrammingExerciseTask> getTasksWithoutInactive(long exerciseId) {
```
Comment (b-fein) [inline src/main/java/de/tum/in/www1/artemis/service/hestia/ProgrammingExerciseTaskService.java:129]:
```suggestion
    public Set<ProgrammingExerciseTask> getTasksWithoutInactiveTestCases(long exerciseId) {
```
(nitpick) It’s not the tasks themself that are inactive, the inactive test cases associated with the task are filtered.

File: src/test/java/de/tum/in/www1/artemis/hestia/ProgrammingExerciseTaskServiceTest.java [hunk 2]
```diff
@@ -1,7 +1,9 @@
 package de.tum.in.www1.artemis.hestia;
 
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.in;
 
+import java.util.Arrays;
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/hestia/ProgrammingExerciseTaskServiceTest.java:6]: unused?

File: src/test/java/de/tum/in/www1/artemis/hestia/ProgrammingExerciseTaskServiceTest.java [hunk 3]
```diff
@@ -173,6 +175,44 @@ void testDeleteWithCodeHints() {
...
     }
 
+    @Test
+    @WithMockUser(username = "instructor1", roles = "INSTRUCTOR")
+    void getTasksWithoutInactiveFiltersOutInactive() {
+        programmingExercise = programmingExerciseRepository
+            .findByIdWithEagerTestCasesStaticCodeAnalysisCategoriesHintsAndTemplateAndSolutionParticipationsAndAuxRepos(programmingExercise.getId()).orElseThrow();
+        programmingExerciseTestCaseRepository.deleteAll(programmingExercise.getTestCases());
+
+        String[] testCaseNames = new String[] { "testClass[BubbleSort]", "testParametrized(Parameter1, 2)[1]" };
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/hestia/ProgrammingExerciseTaskServiceTest.java:185]:
```suggestion
        String[] testCaseNames = { "testClass[BubbleSort]", "testParametrized(Parameter1, 2)[1]" };
```

File: src/test/java/de/tum/in/www1/artemis/hestia/ProgrammingExerciseTaskServiceTest.java [hunk 4]
```diff
@@ -173,6 +175,44 @@ void testDeleteWithCodeHints() {
...
+            var testCase = new ProgrammingExerciseTestCase();
+            testCase.setExercise(programmingExercise);
+            testCase.setTestName(name);
+            testCase.setActive(true);
+            programmingExerciseTestCaseRepository.save(testCase);
+        }
+
+        var testCase = new ProgrammingExerciseTestCase();
+        testCase.setExercise(programmingExercise);
+        testCase.setTestName( "testWithBraces()");
```
Comment (b-fein) [inline src/test/java/de/tum/in/www1/artemis/hestia/ProgrammingExerciseTaskServiceTest.java:196]:
```suggestion
        testCase.setTestName("testWithBraces()");
```

**Review by b-fein — APPROVED** [review]
Code looks good with only minor optional suggestions.
Automatic code formatting should run on the changed code.

File: src/test/java/de/tum/in/www1/artemis/hestia/ProgrammingExerciseTaskServiceTest.java (hunk 2 above)
Comment (simon-hng) [reply src/test/java/de/tum/in/www1/artemis/hestia/ProgrammingExerciseTaskServiceTest.java:6]: yes

**Review by b-fein — APPROVED** [review]
Code changes since last review look good.

**Review by DominikRemo — APPROVED** [review]
Reapproved

**Review by frederik-kt — APPROVED** [review]
Latest code changes look good.

**Review by rriyaldhi — APPROVED** [review]
Reapprove after refactoring

**Comment by bassner** [comment]
@simon-hng Can we get a test coverage table, please?

**Comment by simon-hng** [comment]
> @simon-hng Can we get a test coverage table, please?

I added the coverage table

**Pull request merged**